client/state/test/utils.js

```diff
@@ -235,4 +235,51 @@ describe( 'utils', () => {
 			expect( state ).to.eql( [ 0, 1 ] );
 		} );
 	} );
+
+	describe( 'createSiteReducer()', () => {
+		before( () => {
+			reducer = createSiteReducer( [], {
+				TEST_ADD: ( state, { value } ) => includes( state, value ) ? state : [ ...state, value ]
+			}, {
+				type: 'array',
+				items: {
+					type: 'number'
+				}
+			} );
+		} );
+
+		it( 'should extend createReducer, keying by action.siteId', () => {
+			const state = reducer( undefined, {
+				type: 'TEST_ADD',
+				siteId: 2916284,
+				value: 1
+			} );
+
+			expect( state ).to.eql( { 2916284: [ 1 ] } );
+		} );
+
+		it( 'should optimize returned state as same reference if no change', () => {
+			const originalState = reducer( deepFreeze( {} ), {
+				type: 'TEST_ADD',
+				siteId: 2916284,
+				value: 1
+			} );
+			const state = reducer( originalState, {
+				type: 'TEST_ADD',
+				siteId: 2916284,
+				value: 1
+			} );
+
+			expect( state ).to.equal( originalState );
+		} );
+
+		it( 'should augment schema to account for site keyed structure', () => {
```

Comment:
Optional, but I thought it might be interesting to have these cases:
1. A test to verify that the siteId `patternProperties` works
2. A test to verify that the item schema is being respected. (For example passing through an object instead of an array).